src/spatialmath/utils.rs

```diff
@@ -45,42 +48,394 @@ impl EulerAngles {
+    new_x: Quaternion<f64>, new_z: Quaternion<f64>
+) -> f64 {
+    if 1.0 - new_z.k.abs() > ANGLE_ACCEPTANCE {
+        let new_z_imag = new_z.imag();
+        let new_x_imag = new_x.imag();
+        let z_imag_axis = Vector3::z_axis();
+
+        let normal_1 = new_z_imag.cross(&new_x_imag);
+        let normal_2 = new_z_imag.cross(&z_imag_axis);
+        let cos_theta_cand = normal_1.dot(&normal_2) / (normal_1.norm() * normal_2.norm());
+        let cos_theta = match cos_theta_cand {
+            val if val < -1.0 => -1.0,
+            val if val > 1.0 => 1.0,
+            _ => cos_theta_cand
+        };
+
+        return match cos_theta.acos() {
+            val if val > ANGLE_ACCEPTANCE => {
+                let new_z_imag_unit = UnitVector3::new_normalize(new_z_imag);
+                let rot_quat_unit = UnitQuaternion::from_axis_angle(&new_z_imag_unit, -1.0 * val);
+                let rot_quat = rot_quat_unit.quaternion();
+                let z_axis_quat = Quaternion::new(0.0, 0.0, 0.0, 1.0);
+                let test_z = (rot_quat * z_axis_quat) * rot_quat.conjugate();
+                let test_z_imag = test_z.imag();
+
+                let normal_3 = new_z_imag.cross(&test_z_imag);
+                let cos_test = normal_1.dot(&normal_3) / (normal_3.norm() * normal_1.norm());
+                match cos_test {
+                    val2 if (1.0 - val2) < (ANGLE_ACCEPTANCE * ANGLE_ACCEPTANCE) => -1.0 * val,
+                    _ => val
+                }
+            },
+            _ => 0.0
+        };
+    }
+    match new_z.k {
+        val if val < 0.0 => -1.0 * new_x.j.atan2(new_x.i),
+        _ => -1.0 * new_x.j.atan2(new_x.i * -1.0)
+    }
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+pub struct OrientationVector {
+    pub o_vector: UnitVector3<f64>,
+    pub theta: f64
+}
+
+impl OrientationVector {
+    pub fn new(o_x: f64, o_y: f64, o_z: f64, theta: f64) -> Self {
+        let o_vector = UnitVector3::new_normalize(Vector3::new(o_x, o_y, o_z));
+        OrientationVector{o_vector, theta}
+    }
+
+    pub fn to_quaternion(&self) -> Quaternion<f64> {
+        let lat = self.o_vector.z.acos();
+        let lon = match self.o_vector.z {
+            val if 1.0 - val > ANGLE_ACCEPTANCE => self.o_vector.y.atan2(self.o_vector.x),
+            _ => 0.0
+        };
+
+        // convert angles as euler angles (lon, lat, theta) to quaternion
+        // using the zyz rotational order
+        let s: [f64;3] = [
+            (lon / 2.0).sin(),
+            (lat / 2.0).sin(),
+            (self.theta / 2.0).sin()
+        ];
+
+        let c: [f64;3] = [
+            (lon / 2.0).cos(),
+            (lat / 2.0).cos(),
+            (self.theta / 2.0).cos()
+        ];
+
+        let real = c[0]*c[1]*c[2] - s[0]*c[1]*s[2];
+        let i = c[0]*s[1]*s[2] - s[0]*s[1]*c[2];
+        let j = c[0]*s[1]*c[2] + s[0]*s[1]*s[2];
+        let k = s[0]*c[1]*c[2] + c[0]*c[1]*s[2];
+
+        Quaternion::new(real, i, j, k)
```

Comment:
Not familiar with Rust, but would it be faster to put this math directly in the `new()` call and not initialize the variables?

Reply:
I don't think so, I think Rust ownership rules make both ways essentially equivalent